drodrpg/DRODLib/DbSavedGames.cpp

```diff
@@ -3113,6 +3119,129 @@ void CDbSavedGames::UpdatePlayerTallies(
 	}
 }
 
+//******************************************************************************
+RoomMapStates CDbSavedGames::LoadMapStates(const UINT dwPlayerID, const CIDSet& rooms)
+//Retrieve all states from ST_TotalMapStates for the supplied rooms
+{
+	RoomMapStates mapStates;
+
+	const UINT dwID = FindByType(ST_TotalMapStates, dwPlayerID);
+	CDbSavedGame* pMapStateSave = GetByID(dwID);
+	if (!pMapStateSave)
+		return mapStates;
+
+	CIDSet exploredRooms = pMapStateSave->GetExploredRooms(true, true);
+	exploredRooms.intersect(rooms);
+	CIDSet::const_iterator room;
+	for (room = exploredRooms.begin(); room != exploredRooms.end(); ++room)
+	{
+		UINT roomID = *room;
+		vector<ExploredRoom*>::const_iterator room;
+		for (room = pMapStateSave->ExploredRooms.begin(); room != pMapStateSave->ExploredRooms.end(); ++room)
+		{
+			ExploredRoom* pRoom = *room;
```

Comment:
You can double-dereference pointers for greater efficiency on repeat accesses, like this:
`const ExploredRoom& room = **room;`
and then access the values directly like
`if (room.roomID == roomID)`, etc.